docs/architecture/engine.md

```diff
@@ -0,0 +1,197 @@
+
+1. Does the series exist? and if it does
+2. Does the data exist in an "active buffer" (actively being compressed by an encoder), cached in-memory, on disk, or some combination of all three?
+
+Determining whether the series exists is simple. M3DB looks up the series in the shard object. If it exists, then the series exists. If it doesn't, then M3DB consults an in-memory bloom filter(s) for that shard / block start combination(s) to determine if the series exists on disk.
+
+If the series exists, then for every block that the request spans, M3DB needs to consolidate data from the active buffers, in-memory cache, and fileset files (disk).
+
+Lets imagine a read for a given series that requests the last 6 hours worth of data, and an M3DB namespace that is configured with a blocksize of 2 hours (I.E we need to find 3 different blocks.)
+
+If the current time is 8PM, then the location of the requested blocks might be as follows:
+
+```
+[2PM - 4PM (Fileset file)] - Sealed and flushed block that isn't cached
+[4PM - 6PM (In-memory cache)] - Sealed and flush block that is cached
+[6PM - 8PM (active buffer)] - Hasn't been sealed or flushed yet
+```
+
+Then M3DB will need to consolidate:
+
+1) The not-yet-sealed block from the active buffers / encoders (located inside an internal lookup in the Series object) **[6PM - 8PM]**
+2) The in-memory cached block (also located inside an internal lookup in the Series object) **[4PM - 6PM]**
+3) The block from disk (the block retrieve from disk will then be cached according to the current [caching policy](caching.md) **[2PM - 4PM]**
+
+Retrieving blocks from the active buffers and in-memory cache is simple, the data is already present in memory and easily accessible via hashmaps keyed by series ID. Retrieving a block from disk is more complicated. The flow for retrieving a block from disk is as follows:
+
+1. Consult the in-memory bloom filter to determine if its possible the series exists on disk.
+2. If the bloom filter returns positive, then binary search the in-memory index summaries to find the nearest index entry that is *before* the series ID that we're searching for. Review the `index_lookup.go` file for implementation details.
+3. Jump to the offset in the index file that we obtained from the binary search in the previous step, and begin scanning forward until we identify the index entry for the series ID we're looking for *or* we get far enough in the index file that it becomes clear that the ID we're looking for doesn't exist (this is possible because the index file is sorted by ID)
+4. Jump to the offset in the data file that we obtained from scanning the index file in the previous step, and begin streaming data.
+
+
+Once M3DB has retrieved the three blocks from their respective locations in memory / on-disk, it will transmit all of the data back to the client. 
+
```

Comment:
You need to further describe the consistency guarantees the client will check before returning success/failure.

Reply:
done